### PR TITLE
Refactor `golden_tests_harvester`, throw when not `--dry-run`, add tests.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/144948.

I went a little farther, as I'm not looking forward to debugging this in the future without tests or examples.

### CODE REVIEW

File: tools/golden_tests_harvester/lib/golden_tests_harvester.dart [hunk 1]
```diff
@@ -2,42 +2,98 @@
...
+///       "maxDiffPixelsPercent": 0.01,
+///
+///       // Called `pixelColorDelta` in Skia Gold.
+///       "maxColorDelta": 0
+///     }
+///   ]
+/// }
+/// ```
+///
+/// If [dryRun] is true, the images will _not_ be uploaded to Skia Gold, and
```
Comment (gaaclarke) [inline tools/golden_tests_harvester/lib/golden_tests_harvester.dart:51]: `dryRun` is not apropos to `harvest`, I think you refactored this and didn't yet update the docstring.

File: tools/golden_tests_harvester/lib/src/digests_json_format.dart [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+///
+/// A single file is typically used to represent the results of a test suite,
+/// or a series of tests that have been run with the same [dimensions]. For
+/// example, "Impeller Unittests on Vulkan running on MacOS" might generate
+/// a `digests.json` file.
+///
+/// Other tools (perhaps implemented in other languages, like C++) can use this
+/// format to communicate with the `golden_tests_harvester` tool without
+/// relying on the tool directly (or the Dart SDK).
+final class Digests {
```
Comment (gaaclarke) [inline tools/golden_tests_harvester/lib/src/digests_json_format.dart:17]: This code seems like the sort of thing we should be automating, can we use the dart json parser to do this?  We would probably need custom builders which is a showstopper, eh?

File: tools/golden_tests_harvester/test/golden_tests_harvester_test.dart [hunk 3]
```diff
@@ -2,4 +2,147 @@
...
+      await callback(tempDirectory);
+    } finally {
+      await tempDirectory.delete(recursive: true);
+    }
+  }
+
+  test('should require a file named "digests.json" in the working directory', () async {
+    await withTempDirectory((io.Directory tempDirectory) async {
+      final StringSink stderr = StringBuffer();
+      try {
```
Comment (gaaclarke) [inline tools/golden_tests_harvester/test/golden_tests_harvester_test.dart:25]: You use this pattern a few times.  You could factor this out to 
```dart
expectThrow<io.FileSystemException>((){ await harvest(...);}, (io.FileExceptionException e) {
  expect(e.path, contains('digest.json'));
});
```

**Review by gaaclarke — APPROVED** [review]
LGTM, just a few notes

File: tools/golden_tests_harvester/lib/golden_tests_harvester.dart (hunk 1 above)
Comment (matanlurey) [reply tools/golden_tests_harvester/lib/golden_tests_harvester.dart]: Right, thanks for catching the refactor!

File: tools/golden_tests_harvester/README.md [hunk 4]
```diff
@@ -1,14 +1,31 @@
 # Golden Tests Harvester
 
-Reaps the output of impeller's golden image tests and sends it to Skia gold.
+A command-line tool that uploads golden images to Skia gold from a directory.
 
 ## Usage
 
+This program assumes you've _already run_ a suite of golden tests that produce
+a directory of images, of which the format, a JSON file named `digests.json`, is
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/golden_tests_harvester/README.md]: Done!

File: tools/golden_tests_harvester/lib/src/digests_json_format.dart (hunk 2 above)
Comment (matanlurey) [reply tools/golden_tests_harvester/lib/src/digests_json_format.dart:17]: I am guessing (?) this format will virtually never change, or if it does, we'll catch it in a unit test.

I am not opposed to generating this, but it wouldn't be a priority for me.

File: tools/golden_tests_harvester/test/golden_tests_harvester_test.dart (hunk 3 above)
Comment (matanlurey) [reply tools/golden_tests_harvester/test/golden_tests_harvester_test.dart]: Reasonable suggestion, done!

**Pull request merged**